### PR TITLE
Tile and coordinate manipulation library

### PR DESCRIPTION
**Pull request by daviddrysdale**
I'm not necessarily expecting to merge this, as it probably duplicates logic that already exists in various places.  (However, it was useful for getting my head back around Merkle tree proofs, coordinates and tiles in preparation for future work.)

### CODE REVIEW

**Comment by daviddrysdale** [comment]
Might be of passing interest to @FiloSottile and/or @katiehockman

**Comment by Martin2112** [comment]
Could go in the examples repo maybe?

File: merkle/tile/coords.go [hunk 1]
```diff
@@ -0,0 +1,396 @@
...
+// required sibling leaf and ending with a child of the root node, following the
+// algorithm from [RFC6962] section 2.1.1.
+func InclusionPath(idx, treeSize uint64) []MerkleCoords {
+	if treeSize < 2 {
+		// If there's no leaf or a single leaf, no inclusion path is needed.
+		glog.V(4).Infof("PATH(%d, D[%d]) = nil", idx, treeSize)
+		return nil
+	}
+	if idx >= treeSize {
+		return nil // OUTATREE
```
Comment (Martin2112) [inline merkle/tile/coords.go]: If you return an empty slice above it will be distinguishable.

File: merkle/tile/store.go [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+}
+
+func (e MissingTileError) Error() string {
+	return fmt.Sprintf("tile %v not available", e.Coords)
+}
+
+// RequiredTilesError indicates that hash retrieval failed because specific
+// tiles were missing or incomplete
+type RequiredTilesError struct {
+	Missing    []MissingTileError
```
Comment (Martin2112) [inline merkle/tile/store.go]: This is some elaborate error reporting. You'll be writing Java next.

File: merkle/tile/store.go [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+// type RequiredTilesError.
+func ConsistencyProofFromStore(store Store, fromSize, toSize uint64) ([][]byte, error) {
+	return proofFromPath(store, ConsistencyPath(fromSize, toSize))
+}
+
+// MemoryStore is a Store implementation that just holds tile contents in
+// memory, mainly for testing.
+type MemoryStore struct {
+	tileHeight   uint64
+	hashChildren func(l, r []byte) []byte
```
Comment (Martin2112) [inline merkle/tile/store.go]: Could use the `hashers.LogHasher` type but it does come with baggage.

File: merkle/tile/tiles.go [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+// Package tile includes functionality for describing a Merkle tree in
+// terms of tiles and coordinates.
+package tile
+
+import (
+	"bytes"
+	"fmt"
+)
+
+// Coords represents the location of a single tile from a cover of a Merkle tree.
```
Comment (Martin2112) [inline merkle/tile/tiles.go]: Might help to define "a cover" here? Also further down "encompasses" is used. Possibly keeping the same terms will help readers?

File: merkle/tile/tiles.go [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+// RootHash returns the Merkle tree hash that encompasses a complete tile (and
+// so is associated with the coordinates t.Above()).  Fails for incomplete tiles.
+func (t *Tile) RootHash() ([]byte, error) {
+	if !t.IsComplete() {
+		return nil, fmt.Errorf("root hash unavailable for incomplete tile %v", t)
+	}
+	top := t.Top()
+	l, err := t.HashFor(top[0])
+	if err != nil {
+		// Unhittable as we've already checked for a complete tile.
```
Comment (Martin2112) [inline merkle/tile/tiles.go]: Maybe "Unexpected" here and the next one?

**Comment by AlCutter** [comment]
This is a nice resource to keep around as a branch for didactic reasons, but we're unlikely to merge into the main branch so I'll close this PR for now as we're having a tidy up.

**Pull request closed without merging**